File: android/HelloWorld/DecodeFromAnImage/src/main/java/com/dynamsoft/dbr/decodefromanimage/FileUtil.java
package com.dynamsoft.dbr.decodefromanimage;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;

public class FileUtil {
    public static Bitmap fileBytesToBitmap(byte[] fileBytes) {
        if (fileBytes != null) {
            BitmapFactory.Options opts = new BitmapFactory.Options();
            opts.inJustDecodeBounds = true;
            BitmapFactory.decodeByteArray(fileBytes, 0, fileBytes.length, opts);
            if (opts.outHeight * opts.outWidth > 1920 * 1080) {
                float scale = (float) Math.sqrt((double) (opts.outHeight * opts.outWidth) / (1920 * 1080));
                opts.inTargetDensity = 1;
                opts.inDensity = (int) Math.ceil(scale);
            }
            opts.inJustDecodeBounds = false;
            return BitmapFactory.decodeByteArray(fileBytes, 0, fileBytes.length, opts);
        } else {
            return null;
        }
    }

    public static byte[] inputStreamToBytes(InputStream is) {
        try {
            ByteArrayOutputStream os = new ByteArrayOutputStream();
            byte[] buffer = new byte[1024];
            int len;
            while ((len = is.read(buffer)) != -1) {
                os.write(buffer, 0, len);
            }
            os.close();
            is.close();
            return os.toByteArray();
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static byte[] assetsFileToBytes(Context context, String fileName) {
        try {
            InputStream inputStream = context.getAssets().open(fileName);
            return inputStreamToBytes(inputStream);
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: android/HelloWorld/DecodeFromAnImageKt/src/main/java/com/dynamsoft/decodefromanimagekt/FileUtil.kt
package com.dynamsoft.decodefromanimagekt

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import kotlin.math.ceil

/**
 * @author: dynamsoft
 * Time: 2023/8/25
 * Description:
 */

object FileUtil {
    fun fileBytesToBitmap(fileBytes: ByteArray?): Bitmap? {
        return if (fileBytes != null) {
            val opts = BitmapFactory.Options()
            opts.inJustDecodeBounds = true
            BitmapFactory.decodeByteArray(fileBytes, 0, fileBytes.size, opts)
            if (opts.outHeight * opts.outWidth > 1920 * 1080) {
                val scale =
                    Math.sqrt((opts.outHeight * opts.outWidth).toDouble() / (1920 * 1080)).toFloat()
                opts.inTargetDensity = 1
                opts.inDensity = ceil(scale).toInt()
            }
            opts.inJustDecodeBounds = false
            BitmapFactory.decodeByteArray(fileBytes, 0, fileBytes.size, opts)
        } else {
            null
        }
    }

    fun inputStreamToBytes(`is`: InputStream): ByteArray? {
        return try {
            val os = ByteArrayOutputStream()
            val buffer = ByteArray(1024)
            var len: Int
            while (`is`.read(buffer).also { len = it } != -1) {
                os.write(buffer, 0, len)
            }
            os.close()
            `is`.close()
            os.toByteArray()
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }
    }

    fun assetsFileToBytes(context: Context, fileName: String?): ByteArray? {
        return try {
            val inputStream = context.assets.open(fileName!!)
            inputStreamToBytes(inputStream)
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }
    }
}
